dbms/src/Storages/Transaction/FastAddPeer.cpp

```diff
@@ -0,0 +1,478 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/FailPoint.h>
+#include <Encryption/PosixRandomAccessFile.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/SharedContexts/Disagg.h>
+#include <Storages/Page/V3/CheckpointFile/CPManifestFileReader.h>
+#include <Storages/Page/V3/PageDirectory.h>
+#include <Storages/Page/V3/Universal/RaftDataReader.h>
+#include <Storages/Page/V3/Universal/UniversalPageIdFormatImpl.h>
+#include <Storages/Page/V3/Universal/UniversalPageStorage.h>
+#include <Storages/Page/V3/Universal/UniversalWriteBatchImpl.h>
+#include <Storages/S3/CheckpointManifestS3Set.h>
+#include <Storages/S3/S3Common.h>
+#include <Storages/S3/S3Filename.h>
+#include <Storages/S3/S3GCManager.h>
+#include <Storages/S3/S3RandomAccessFile.h>
+#include <Storages/Transaction/CheckpointInfo.h>
+#include <Storages/Transaction/FastAddPeerAsyncTasksImpl.h>
+#include <Storages/Transaction/KVStore.h>
+#include <Storages/Transaction/ProxyFFI.h>
+#include <Storages/Transaction/ProxyFFICommon.h>
+#include <Storages/Transaction/Region.h>
+#include <Storages/Transaction/TMTContext.h>
+#include <fmt/core.h>
+
+namespace DB
+{
+
+FastAddPeerContext::FastAddPeerContext(uint64_t thread_count)
+{
+    if (thread_count == 0)
+    {
+        static constexpr int ffi_handle_sec = 5;
+        static constexpr int region_per_sec = 2;
+        thread_count = ffi_handle_sec * region_per_sec;
+    }
+    tasks_trace = std::make_shared<AsyncTasks>(thread_count);
+}
+
+TempUniversalPageStoragePtr FastAddPeerContext::getTempUniversalPageStorage(UInt64 store_id, UInt64 upload_seq)
+{
+    std::unique_lock lock(ps_cache_mu);
+    auto iter = temp_ps_cache.find(store_id);
+    if (iter != temp_ps_cache.end() && iter->second.first >= upload_seq)
+    {
+        return iter->second.second;
+    }
+    return nullptr;
+}
+
+void FastAddPeerContext::updateTempUniversalPageStorage(UInt64 store_id, UInt64 upload_seq, TempUniversalPageStoragePtr temp_ps)
+{
+    std::unique_lock lock(ps_cache_mu);
+    auto iter = temp_ps_cache.find(store_id);
+    if (iter != temp_ps_cache.end() && iter->second.first >= upload_seq)
+        return;
+
+    temp_ps_cache.erase(store_id);
+    temp_ps_cache.emplace(store_id, std::make_pair(upload_seq, temp_ps));
+}
+
+void FastAddPeerContext::insertSegmentEndKeyInfoToCache(NamespaceId table_id, const DM::RowKeyValue & key, UInt64 segment_id)
+{
+    std::unique_lock lock(range_cache_mu);
+    auto & end_key_to_id_map = segment_range_cache[table_id];
+    end_key_to_id_map.erase(key);
```

Comment:
Why do we need to erase first?

Reply:
Just want to make sure later `emplace` success. Use a different way now.